### PR TITLE
Use LoadBalancer Service

### PR DESCRIPTION
**Pull request by DanielArndt**
# Description

Exposes an external LoadBalancer service for the UPF, similar to https://github.com/canonical/sdcore-amf-operator/pull/24 for the AMF

# Checklist:

- [x] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that validate the behaviour of the software
- [x] I validated that new and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [ ] I have bumped the version of the library

### CODE REVIEW

File: src/charm.py
Comment (gatici) [inline src/charm.py:126]: This will create a service with a port assigned by Loadbalancer. 
service/upf-external                          LoadBalancer   10.152.183.224   10.0.0.2      8805:31523/UDP      2m
I have a question here.
Do we need to allow to define any specific port by user ? At the moment, it will be automatically assigned by Loadbalancer service. (31523)

Second issue:
I am not sure, if we need to consider this at the moment but for public cloud environments, we may need to provide Loadbalancer IP, if cloud-provider requires to specify the loadBalancerIP.

**Review by Gmerold — CHANGES_REQUESTED** [review]
Two general comments:
1. I think we could mention in the README that it is possible to expose UPF, but LB is required for that. We could also have an integration test validating the exposure. WDYT?
2. Please add docstrings :)

File: src/charm.py
Comment (gatici) [inline src/charm.py:124]: I suggest to create a variables for pfcp port name and value (8805).

Comment (DanielArndt) [reply src/charm.py:124]: I believe the intention was to have both the internal and the external one. I'd have to rely on @ghislainbourgeois here to explain the benefit of doing it that way.

**Comment by DanielArndt** [comment]
> Two general comments:
> 
> 1. I think we could mention in the README that it is possible to expose UPF, but LB is required for that. We could also have an integration test validating the exposure. WDYT?

Sure, I can add that to the README. 

I'd be happy to give integration tests a shot, but at this point I'm not quite sure what that would look like. I'll have to tackle those when I get back on Wednesday.

> 2. Please add docstrings :)

Happy to add docstrings where you think they would be useful. I'm not a fan of docstrings on internal functions where the docstring just reiterates the function name. i.e.

```
def _create_external_upf_service(self) -> None:
    """Creates the external UPF service."""
````

The above seems like a waste, but maybe there is something you'd like me to be more explicit about?

File: src/charm.py
Comment (DanielArndt) [inline src/charm.py:107]: Could be a staticmethod

File: src/charm.py
Comment (DanielArndt) [inline src/charm.py:136]: Could be a staticmethod

File: src/charm.py
Comment (Gmerold) [reply src/charm.py:124]: I don't think we need both. The main goal of this story is to expose the pfcp service through the load balancer, to be able to deploy UPF separately from the rest of the 5G core.

File: src/charm.py
Comment (Gmerold) [reply src/charm.py:126]: Ad. 1. I don't think it's a problem. What matters is that we need our service to be available on port `8805`. Which internal port it maps to doesn't really matter.
Ad. 2. I'm not sure I understand the problem. Would you mind giving an example?

File: src/charm.py
Comment (DanielArndt) [reply src/charm.py:124]: More than happy to remove it, but @ghislainbourgeois was the one who suggested leaving it, so I just want to make sure he didn't have something else in mind. I'll try to grab his attention.

File: src/charm.py
Comment (DanielArndt) [reply src/charm.py:107]: Oops, no it can't.

File: src/charm.py
Comment (Gmerold) [reply src/charm.py:127]: OK, that explains it :) I didn't know that option ever existed :D

**Comment by gatici** [comment]
We need to rebase it on top of main.

**Review by ghislainbourgeois — APPROVED** [review]

File: src/charm.py
Comment (Gmerold) [reply src/charm.py]: Either way is fine for me. I really don't expect too many users of our tutorial at this point anyway ;) 
My only point here was that we probably don't want to keep both.
@DanielArndt I leave it to you to decide which approach you prefer.

**Review by Gmerold — APPROVED** [review]

**Review by gatici — APPROVED** [review]
Needs to be rebased before merging. LGTM.

File: src/charm.py
Comment (DanielArndt) [reply src/charm.py]: I'd prefer the incremental approach, so I've made follow-up tasks for this.

**Pull request merged**
